Moved filter query building logic and added meta classes to handle integrity of serializers

sampeka opened the pull request:
I would like to build in some greater support for preprocess filtering without using the `FacetFilter`, but the query building logic and the filters were quite tightly coupled. So I've refactored some of the codebase to make it easier to change things.

Changes:
* Moved the parameter-building logic for filters (with the exception of GeoFilter since the tests were being skipped) to `QueryBuilder`. This should make it easier to subclass and build more specific backend support without creating lots of filters.
* I've added some meta classes in the serializers so that you don't need to use `getattr` for the meta attributes.
* Now if meta classes are incorrectly configured (or a serializer does not define Meta without being abstract) it will throw an exception when the class itself is defined, rather than the method called. I think this is a good change but has a problem with backwards compatibility? 
* Raised test settings log levels (#40)

If you're not happy with any of these changes, let me know :) I've added test coverage for the meta classes, but haven't bothered with docs unless you're happy with this.

rhblind commented:
Hi,
Sounds great!

I'll have a look at it during the weekend or so. I'm quite occupied with some work these days so I doubt I'll have a chance before =)

sampeka commented:
Great, and no problem - whenever you have some spare time :) I doubt travis will pass first time anyway so I'll try and sort that out by the weekend

rhblind commented:
Hi,
There's a lot of good ideas here but I need it to pass the current test suite in order to merge.
However, maybe this kind of refactoring is better suited for a 1.6 release, as it might include breaking API changes/backwards incompatibility.

I have been thinking about the 1.6 release for a while and it includes some other stuff I'd like to get cleaned up as well.

What do you think?

sampeka commented:
Sorry for the delayed response, that sounds good to me. I've just fixed the python2 issue (hopefully) and rebased on master. 

I think that Django 1.5 will still fail for python 2 and 3, seems to be a slight difference in how the six package handles it's meta class stuff. How important is it to support Django 1.5? Haystack aren't supporting it and it's not maintained anymore, but I could look into it more if you'd like?

rhblind commented:
Hi, good stuff.
I've already started a releae branch `release/v1.6.0` which you can check out.

I think this would be a good time to drop support for old versions and start following the [Haystack supported versions](https://github.com/django-haystack/django-haystack#requirements).

sampeka commented:
Agreed :)

Travis failed again, I'll stop being lazy now and fix it properly now ;) When I get this passing do you want me to make the PR against 1.6.0 instead? Looks like I'll need to pull in some of your changes anyway.

rhblind commented:
Good.

Yes, please, that would've been great!
